**MAESTRÍA EN INTELIGENCIA ARTIFICIAL APLICADA**

**Curso: TC5053 - Ciencia y analítica de datos**

Tecnológico de Monterrey

Prof Grettel Barceló Alonso

**Semana 3**
Bases, almacenes y manipulación de datos

---

*   NOMBRE: Elda Estefanía Lara Cruz
*   MATRÍCULA: A01747420


---

En esta actividad usarás una base de datos relacional basada en el informe de participación y la lista del top 10 de Netflix. Incluye películas y programas de televisión, así como información sobre temporadas, métricas de visualización, fechas de estreno, duración y más, organizada en las siguientes tablas:

* `movie`: Información general de las películas.

* `tv_show`: Información general de los programas de televisión.

* `season`: Datos de las temporadas asociadas a cada programa de TV.

* `view_summary`: Métricas de visualización y rendimiento de películas o temporadas.

Revisa con detalle su esquema para que comprendas cómo se relacionan las tablas anteriores.

**NOTA IMPORTANTE:** Asegúrate de responder *explícitamente* todos los cuestionamientos.


`PyMySQL` es una librería escrita en Python puro que funciona como conector (*driver*) para motores de bases de datos MySQL, permitiendo abrir conexiones, ejecutar consultas SQL y recuperar resultados directamente desde programas en Python.

In [32]:
pip install pymysql

`SQLAlchemy` es una librería de Python que facilita la interacción con bases de datos y permite mantener un pool de conexiones eficiente, gestionar *commits* y *rollbacks* de forma automática y asegurar que múltiples conexiones simultáneas se manejen de manera segura, incluso cuando se ejecutan consultas SQL “puras”

In [33]:
# Importa las librerías necesarias
import pymysql
import sqlalchemy as sqla
import pandas as pd

Se crea una conexión (`conn`) para luego invocar declaraciones SQL.

In [34]:
# motor+driver://usuarioBD:clave@ipHostDBMS:puerto/esquema
# pool_recycle controla el tiempo máximo de vida de una conexión en el pool (3600 segundos = 1 hora)
db = sqla.create_engine('mysql+pymysql://mnaTC5053User:mnaTC5053Pass!@20.127.169.219:3306/Netflix', pool_recycle=3600)
conn = db.connect()

Para que tus consultas sean más legibles y fáciles de mantener, puedes usar este formato multilínea con comillas triples y `sqla.text()`. Por ejemplo:

```
query = sqla.text("""
  SELECT ---
  FROM ---
  WHERE ---
""")
pd.read_sql(query, conn)
```

1.	Extrae toda la información de las películas que duran más de 5 horas.

> Add blockquote



In [35]:
query = sqla.text("""
  SELECT *
  FROM movie
  WHERE runtime > 300
""")
pd.read_sql(query, conn)

,id,created_date,modified_date,available_globally,locale,original_title,release_date,runtime,title
0,5793,2024-01-01,2024-01-01,b'\x00',None,日本統一シリーズ: 映画シリーズ,None,3892,Nihontouitsu Series: Film Series
1,5794,2024-01-01,2024-01-01,b'\x00',None,釣りバカ日誌: 映画シリーズ,None,2120,Free and Easy Series: Film Series
2,5874,2024-01-01,2024-01-01,b'\x01',None,None,2021-08-06,312,Navarasa: Limited Series
3,9729,2024-01-01,2024-01-01,b'\x00',None,織田同志会 織田征仁: 映画シリーズ,None,710,Seiji Oda: Film Series
4,9730,2024-01-01,2024-01-01,b'\x00',None,キングダム～首領になった男～: 映画シリーズ,None,427,Kingdom ~ The Man Who Became the Top ~: Film S...


2.	¿Cuál es el porcentaje de películas disponibles únicamente en EU en relación con el total, excluyendo los valores `NULL`? La consulta SQL debe entregar directamente el porcentaje final. No se deben devolver resultados parciales para realizar el cálculo en Python.

In [36]:
query = sqla.text("""
  SELECT ROUND(100 * SUM(available_globally = 0) / COUNT(*), 2) AS porcentaje_solo_eu
  FROM movie
  WHERE available_globally IS NOT NULL
""")
pd.read_sql(query, conn)

,porcentaje_solo_eu
0,83.57


3.	¿Cuáles son los idiomas o regiones originales en la tabla de películas?
* ¿Cuántos registros tienen el campo `locale` con valor `NULL`? (NULL en SQL ⇔ None en Python)

In [37]:
query = sqla.text("""
  SELECT COUNT(*)                                   AS total_peliculas,
         SUM(locale IS NULL)                        AS locale_null,
         SUM(locale = '')                           AS locale_vacio,
         COUNT(DISTINCT locale)                     AS idiomas_distintos
  FROM movie
""")
pd.read_sql(query, conn)

,total_peliculas,locale_null,locale_vacio,idiomas_distintos
0,11787,11255.0,0.0,1


In [38]:
query = sqla.text("""
  SELECT COUNT(*) AS locale_null
  FROM movie
  WHERE locale IS NULL
""")
pd.read_sql(query, conn)

,locale_null
0,11255


4.	Asumiendo que los valores `NULL` en `locale` corresponden a otro idioma (diferente del inglés), el título original de la película NO debería coincidir con el título principal en dichos registros.
* Determina cuántas películas tienen títulos diferentes en estos dos campos (`title` y `original_title`).
*  ¿Coinciden los resultados (cantidad de `NULL` y títulos diferentes)? Si no es así, identifica qué características tienen los registros restantes.
* Finalmente, concluye si la suposición de que los valores `NULL` en `locale` indican que la película está en otro idioma es válida.

In [39]:
# a) y b) Cantidades a comparar
query = sqla.text("""
  SELECT SUM(locale IS NULL) AS locale_null,
         SUM(original_title IS NOT NULL
             AND BINARY title <> BINARY original_title) AS titulos_diferentes
  FROM movie
""")
pd.read_sql(query, conn)

,locale_null,titulos_diferentes
0,11255.0,3964.0


In [40]:
# c) Cruce completo para identificar los registros que no coinciden
query = sqla.text("""
  SELECT locale IS NULL         AS locale_nulo,
         original_title IS NULL AS original_title_nulo,
         COALESCE(BINARY title <> BINARY original_title, 0) AS titulo_distinto,
         COUNT(*)               AS peliculas
  FROM movie
  GROUP BY 1, 2, 3
  ORDER BY 1, 2, 3
""")
pd.read_sql(query, conn)

,locale_nulo,original_title_nulo,titulo_distinto,peliculas
0,0,1,0,532
1,1,0,1,3964
2,1,1,0,7291


Los resultados no coinciden: hay 11,255 registros con locale NULL, pero solo 3,964 tienen title distinto de original_title. Los 7,291 restantes tienen tanto locale como original_title en NULL, por lo que no hay un título original con el cual compararlos. Además, ninguna película con locale definido (532 registros) tiene un título original distinto.

La suposición es válida solo parcialmente. Todas las películas con un título original distinto tienen locale NULL, y los ejemplos revisados (Recep İvedik 7, Verónica, Fantômas, Tótem) son títulos de otros idiomas. Sin embargo, el recíproco no se cumple: solo el 35.2% de los registros con locale NULL (3,964 de 11,255) muestra evidencia de otro idioma. Por tanto, un locale NULL no basta para concluir que la película está en otro idioma; en la mayoría de los casos simplemente falta información.


5.	Determina el título de la película que ha permanecido más tiempo en el top 10.

In [41]:
query = sqla.text("""
  SELECT m.title, MAX(v.cumulative_weeks_in_top10) AS semanas_en_top10
  FROM movie m
  JOIN view_summary v ON v.movie_id = m.id
  WHERE v.duration = 'WEEKLY'
  GROUP BY m.id, m.title
  HAVING MAX(v.cumulative_weeks_in_top10) = (
      SELECT MAX(cumulative_weeks_in_top10)
      FROM view_summary
      WHERE movie_id IS NOT NULL AND duration = 'WEEKLY'
  )
""")
pd.read_sql(query, conn)

,title,semanas_en_top10
0,The Boss Baby,22


6.	Identifica los 5 programas de TV con mayor cantidad de temporadas.

In [42]:
query = sqla.text("""
  SELECT t.title AS programa, COUNT(s.id) AS temporadas
  FROM tv_show t
  JOIN season s ON s.tv_show_id = t.id
  GROUP BY t.id, t.title
  ORDER BY temporadas DESC, t.title
  LIMIT 5
""")
pd.read_sql(query, conn)

,programa,temporadas
0,Grey's Anatomy,20
1,Naruto Shippuden,20
2,Heartland (2007),17
3,It's Always Sunny in Philadelphia,16
4,NCIS,15


7.	¿Cuáles son los intervalos de fechas de los resúmenes en la tabla `view_summary` de los períodos (`duration`) semestrales?

In [45]:
query = sqla.text("""
  SELECT DISTINCT start_date, end_date
  FROM view_summary
  WHERE duration = 'SEMI_ANNUALLY'
  ORDER BY start_date
""")
pd.read_sql(query, conn)

,start_date,end_date
0,2023-07-01,2023-12-31
1,2024-01-01,2024-06-30


In [46]:
pd.read_sql(sqla.text("SELECT DISTINCT duration FROM view_summary"), conn)

,duration
0,SEMI_ANNUALLY
1,WEEKLY


8.	Ordena las temporadas de *Grey's Anatomy* según la cantidad de vistas registradas en el primer período semestral de 2024.
* ¿Cómo interpretarías los resultados obtenidos?

In [47]:
query = sqla.text("""
  SELECT s.title AS temporada, s.season_number, v.views
  FROM tv_show t
  JOIN season s        ON s.tv_show_id = t.id
  JOIN view_summary v  ON v.season_id = s.id
  WHERE t.title = "Grey's Anatomy"
    AND v.duration = 'SEMI_ANNUALLY'
    AND v.start_date = '2024-01-01'
  ORDER BY v.views DESC
""")
pd.read_sql(query, conn)

,temporada,season_number,views
0,Grey's Anatomy: Season 1,1,3600000
1,Grey's Anatomy: Season 2,2,3100000
2,Grey's Anatomy: Season 3,3,2900000
3,Grey's Anatomy: Season 5,5,2900000
4,Grey's Anatomy: Season 4,4,2900000
5,Grey's Anatomy: Season 6,6,2800000
6,Grey's Anatomy: Season 7,7,2700000
7,Grey's Anatomy: Season 8,8,2600000
8,Grey's Anatomy: Season 9,9,2500000
9,Grey's Anatomy: Season 10,10,2400000


La temporada 1 es la más vista del primer semestre de 2024, con 3.6 millones de vistas, seguida de la temporada 2 con 3.1 millones. A partir de ahí, las vistas disminuyen de forma casi continua, hasta 1.7 millones en las temporadas 17 y 18. Las diez primeras temporadas concentran aproximadamente el 62% de las vistas de la serie en el periodo (28.4 de 45.8 millones), lo que sugiere que la audiencia se concentra en las temporadas iniciales, probablemente por quienes empiezan la serie desde el principio.
Hay dos excepciones al patrón. La temporada 19 (2.0 millones) supera a las temporadas 15 a 18, lo que podría deberse a que es una de las más recientes. Y la temporada 20 registra solo 100,000 vistas, un valor muy inferior al resto que posiblemente refleja una disponibilidad parcial en la plataforma durante el periodo.
Las cifras están redondeadas a cientos de miles, por lo que aparecen empates (temporadas 3, 4 y 5; y temporadas 12, 13, 14 y 19) en los que el orden no es significativo. Además, los datos corresponden a un solo semestre y solo a Netflix, por lo que no permiten concluir sobre la popularidad general de la serie.


9.	Todas las consultas anteriores podrían hacerse también con la lógica relacional implementada en Pandas, que permite replicar la mayoría de las operaciones de SQL. Si los dataframes se han llenado como sigue, resuelve la consulta 8 utilizando las funciones de Pandas.

In [48]:
tv_show = pd.read_sql(sqla.text('SELECT * FROM tv_show'), conn)
season = pd.read_sql(sqla.text('SELECT * FROM season'), conn)
view_summary = pd.read_sql(sqla.text('SELECT * FROM view_summary'), conn)

In [49]:
# 1) Programa: solo Grey's Anatomy
show = (tv_show.loc[tv_show['title'] == "Grey's Anatomy", ['id']]
               .rename(columns={'id': 'tv_show_id'}))

# 2) Temporadas, con nombres de columna claros
temporadas = (season[['id', 'tv_show_id', 'title', 'season_number']]
              .rename(columns={'id': 'season_id', 'title': 'temporada'}))

# 3) Resúmenes del primer semestre de 2024
vistas = view_summary.loc[
    (view_summary['duration'] == 'SEMI_ANNUALLY')
    & (pd.to_datetime(view_summary['start_date']) == '2024-01-01'),
    ['season_id', 'views']
]

# 4) Unir y ordenar
res = (show.merge(temporadas, on='tv_show_id')
           .merge(vistas, on='season_id')
           .sort_values(['views', 'season_number'], ascending=[False, True])
           [['temporada', 'season_number', 'views']]
           .reset_index(drop=True))
res

,temporada,season_number,views
0,Grey's Anatomy: Season 1,1.0,3600000
1,Grey's Anatomy: Season 2,2.0,3100000
2,Grey's Anatomy: Season 3,3.0,2900000
3,Grey's Anatomy: Season 4,4.0,2900000
4,Grey's Anatomy: Season 5,5.0,2900000
5,Grey's Anatomy: Season 6,6.0,2800000
6,Grey's Anatomy: Season 7,7.0,2700000
7,Grey's Anatomy: Season 8,8.0,2600000
8,Grey's Anatomy: Season 9,9.0,2500000
9,Grey's Anatomy: Season 10,10.0,2400000


`MySQL` es un sistema de gestión de bases de datos relacional, pero desde Python también es posible extraer información de bases de datos no relacionales, como `Firestore`, `MongoDB` o `Cassandra`, utilizando conectores o integraciones específicas. Esto permite combinar datos de diferentes fuentes según las necesidades del análisis o la aplicación.

En el siguiente ejercicio usarás un ejemplo con `Firestore` desde Python. Para ello utilizarás los módulos `credentials` y `firestore` de la biblioteca `firebase_admin`.

In [50]:
import firebase_admin
from firebase_admin import credentials
from firebase_admin import firestore

En `Firestore`, a diferencia de `MySQL` donde se utiliza un usuario y contraseña para conectarse, la autenticación se realiza mediante un archivo JSON que almacena las credenciales necesarias para acceder a la base de datos. Este archivo contiene las claves y la información de configuración que permiten a Python establecer la conexión de manera segura.

In [52]:
from google.colab import drive
drive.mount('/content/drive')

ValueError: mount failed

In [ ]:
# Debes descargar el archivo de credenciales "consultancy.json" (disponible en las instrucciones de Canvas) y ubicarte en la carpeta donde lo almacenaste
import os
DIR = "/content/drive/MyDrive/Colab Notebooks/MNA/TC5053 - Ciencia y analítica de datos/Actividad3_BD_Manipulacion"
os.chdir(DIR)

In [ ]:
# consultancy.json almacena la clave privada para autenticar una cuenta y autorizar el acceso a los servicios
# A través de la función Certificate(), se regresa una credencial inicializada, que puedes utilizar para crear una nueva instancia de la aplicación
cred = credentials.Certificate('consultancy.json')
firebase_admin.initialize_app(cred)
db = firestore.client()

10.	Investiga cómo leer la colección `EMPLOYEE` y mostrar su contenido en un dataframe. Asegúrate de incluir el `id` en el resultado

In [ ]:
docs = db.collection('EMPLOYEE').stream()

employee = pd.DataFrame([{'id': d.id, **d.to_dict()} for d in docs])
display(employee)

# Cerrar la app para poder volver a ejecutar initialize_app sin error
firebase_admin.delete_app(firebase_admin.get_app())

In [ ]:
firebase_admin.delete_app(firebase_admin.get_app())

---

**Declaración de uso de IA**

Anthropic. (2026). Claude Sonnet 5.5 [Modelo de lenguaje grande], utilizado para generación de código, depuración y apoyo en la redacción de interpretaciones. https://claude.ai/
---